cpp/src/arrow/io/CMakeLists.txt

```diff
@@ -20,6 +20,7 @@
 
 set(ARROW_IO_LINK_LIBS
   arrow_shared
+  dl
```

Comment:
Ah, ok. I had the problem with missing `dlopen` symbols in `parquet-cpp` as there static and shared test libs where mixed. There it would have been only a workaround to link `dl` but as we explicitly use it here, it's fine.